src/java.base/share/classes/sun/security/ssl/KAKeyDerivation.java

```diff
@@ -106,23 +108,26 @@ private SecretKey t13DeriveKey(String algorithm,
 
             CipherSuite.HashAlg hashAlg = context.negotiatedCipherSuite.hashAlg;
             SSLKeyDerivation kd = context.handshakeKeyDerivation;
-            HKDF hkdf = new HKDF(hashAlg.name);
+            KDF hkdf = KDF.getInstance(Utilities.digest2HKDF(hashAlg.name));
             if (kd == null) {   // No PSK is in use.
-                // If PSK is not in use Early Secret will still be
+                // If PSK is not in use, Early Secret will still be
                 // HKDF-Extract(0, 0).
                 byte[] zeros = new byte[hashAlg.hashLength];
                 SecretKeySpec ikm
                         = new SecretKeySpec(zeros, "TlsPreSharedSecret");
-                SecretKey earlySecret
-                        = hkdf.extract(zeros, ikm, "TlsEarlySecret");
+                SecretKey earlySecret = hkdf.deriveKey("TlsEarlySecret",
+                        HKDFParameterSpec.ofExtract().addSalt(zeros)
+                        .addIKM(ikm).extractOnly());
```

Comment:
Maybe no need for `addSalt(zeros)`. I remember salt is by default zeros for HKDF.

Reply:
Yes, I am on the fence about this. Given the specified value is the same as the default, it can be removed. I kept it there so the new code matches the original code completely. Not much difference either way I think.

Reply:
I like having it there to communicate that is really the intent.

Reply:
Ok.